### PR TITLE
Preparation for #8452: 'jquery' package extraction

### PR DESCRIPTION
**Pull request by klimov-paul**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   | yes
| Fixed issues  | #12588, preparing for #8452

 'jquery' package has been extracted as preparation for `yii2-jquery` repo extraction, which will allow to create requirements free version of Yii2 and make composer-asset-plugin installation to be optional.

### CODE REVIEW

**Comment by rob006** [comment]
Can't we just use jQuery from CDN by default?

**Comment by klimov-paul** [comment]
> Can't we just use jQuery from CDN by default?

It makes no difference. PHP framework should not be bound to any JavaScript framework.
It should be developer choice if he wishes to use 'JQuery', 'ReactJS', 'Angular' or anything else.

**Comment by rob006** [comment]
> It makes no difference.

It fixes main issue - you don't need asset plugin anymore and you can install Yii without any frontend dependencies. 

I agree that framework-agnostic approach is generally good thing, but it looks like overkill and unnecessary complication. If I want write extension which extends `GridView`, which class I should use as a base class?

**Comment by klimov-paul** [comment]
> I agree that framework-agnostic approach is generally good thing, but it looks like overkill and unnecessary complication.

Following your arguments, should not we bind Yii2 to Twitter Bootstrap as well? Why keep `yii2-bootstrap` as a separated repo? Does not it produce an 'overkill and unnecessary complication'?

> If I want write extension which extends GridView, which class I should use as a base class?

It depends on what you want to do: you may use `yii\grid\GridView` and provide your own JS layer for its filter - or do not use any JS layer, but a regular HTML form submit. Or you can extend `yii\jquery\GridView` and inherit related JQuery code with it.

I am fixing the issues, which are opened and which coew team has agreed to be fixed. If you wish to debate on necessity of such changes use the related issues for that: #8452, #11792, #10557, #8709, #9525.
When you are able to convince @WadeShuler, @leandrogehlen, @graphicmist, @fabiocarneiro and others that thier arguments are unfounded I will discard this PR.

**Comment by klimov-paul** [comment]
This PR fixes #12588

**Comment by rob006** [comment]
> Following your arguments [...]

Please, be constructive...

> It depends on what you want to do [...]

My extension is not related to JS at all so I may use any of this class. But if I use `yii\grid\GridView` as a base class people who want use jQuery stuff in `GridView` will not be able to use my widget. If I use `yii\jquery\GridView`, I will force users to download all this frontend dependencies. So basically I need create and maintain 2 extensions instead of 1.

> I am fixing the issues, which are opened and which core team has agreed to be fixed. If you wish to debate on necessity of such changes use the related issues for that: #8452, #11792, #10557, #8709, #9525.

Did you at least read these issues before linking? They are mostly about installing frontend dependencies, which may be solved by using CDN fallback. Some of them even proposing using CDN for jQuery.

**Comment by samdark** [comment]
Hmm... @rob006 has a point about having two separate GridViews...

File: framework/UPGRADE.md [hunk 1]
```diff
@@ -57,6 +57,14 @@ Upgrade from Yii 2.0.x
   providing ability to setup custom mail headers. Make sure your provide implementation for those methods, while
   creating your own mailer solution.
 
+* Classes `yii\widgets\ActiveForm`, `yii\widgets\ActiveField`, `yii\grid\GridView`, `yii\captcha\Captcha`, `yii\web\View`
+  have been refactored to be more generic without including any 'JQuery' support and client-side processing (validation, automatic submit etc.).
+  You should use `yii\jquery\*` package classes instead, to make old code function as before. E.g. use `yii\jquery\ActiveForm` instead
+  of `yii\widgets\ActiveForm`, `yii\jquery\GridView` instead of `yii\grid\GridView` and so on.
+
+* Assets `yii\web\JqueryAsset`, `yii\web\YiiAsset`, `yii\validators\ValidationAsset`, `yii\captcha\CaptchaAsset` have been
```
Comment (samdark) [inline framework/UPGRADE.md]: I was planning extracting Captcha in its own package.

**Comment by dynasource** [comment]
I agree with @rob006 , I'm positive towards disentanglement of parts, but I dont like the overuse of inheritance in the framework. This is not exactly favoring 'composition' over inheritance and limits flexibility.

**Comment by samdark** [comment]
@dynasource any ideas on how to do it better?

**Comment by dynasource** [comment]
like I said, encapsulate the logic in dependencies instead of inheritance. All these JS parts deserve their own classes and should not pollute the inheritance tree.

**Comment by dynasource** [comment]
this is also an example of GridView `has` jQuery instead of GridView `is` jQuery
https://en.wikipedia.org/wiki/Composition_over_inheritance#Benefits

**Comment by samdark** [comment]
I understand the theory. I was asking about its practical application for this particular case.

**Comment by klimov-paul** [comment]
> My extension is not related to JS at all so I may use any of this class. But if I use yii\grid\GridView as a base class people who want use jQuery stuff in GridView will not be able to use my widget.

Same happens for `yii2-bootstrap`: if you extend `yii\widgets\ActiveForm` your extension can not be used for `yii\bootstrap\ActiveForm`. 
If you extend `yii\base\View` it will not be able to substitute `yii\web\View` and so on.
It is rather funny to hear so much concerns about 'polluting the inheritance tree' here.

> Did you at least read these issues before linking? They are mostly about installing frontend dependencies, which may be solved by using CDN fallback. Some of them even proposing using CDN for jQuery.

I can ask you the same question. So:

1) #8452 (Requirements free version of Yii 2) requires `yii2-framework` package be independent of any external library, allowing easy composer installation.

2) #11792 suggets splitting main repo into several ones, proposing packages `Recaptcha`, `Migrations`, `RBAC`, which I consider unfounded. But just as you he likes CDN a lot, ignoring the fact someone may use AnguarJS instead of JQuery.

3) #10557 (Separate widgets from core framework) demands all widgets should be placed under separated repo because of assets usage. See my comment here https://github.com/yiisoft/yii2/issues/10557#issuecomment-178571517

4) #8709 (Yii2 for REST applications) demands exluding 'bower packages' from installation as a redundant.

5) #9525 (Jquery dependency) demands exactly what I say: 

> Do you really think is a good idea to have a asset as dependency of a PHP FRAMEWORK?

All those issues have been merged into single one #8452.

**Comment by klimov-paul** [comment]
> All these JS parts deserve their own classes and should not pollute the inheritance tree.

This PR has a lot of room for improvement.
For example: at the present state there is `ClientValidatorBuilder`, which is a slave object of the `yii\jquery\ActiveFrom`. In theory it can be evolved in something, which will consume all client-side code composition - but it is hard to say. There are many design flaws around form client-side composition - I can not deal with all them on my own.

> GridView has jQuery instead of GridView is jQuery

Client script for `GridView` can be handled using a behavior, which usage for widgets has been introduced at #7435. For the `ActiveForm` it is hard to say.

**Comment by rob006** [comment]
> Same happens for yii2-bootstrap [...]

And this *is* a problem. Having one solution like that does not justify blind follow in this direction.

> I can ask you the same question [...]

So, if all these issues are about **installing** assets, why do you think that using CDN will not solve them?

Note that using jQuery as default  frontend framework does not force you to use jQuery. If we extract registering assets  and JS to separate protected method, changing fronted framework will be pretty easy - you need to override only one method (could be done by trait or behavior). We can even make it configurable. And this will practically solve this issue without create extra non-jQuery classes - you don't need install frontend dependencies, and you don't need use jQuery at all.

**Comment by klimov-paul** [comment]
> Note that using jQuery as default frontend framework does not force you to use jQuery. If we extract registering assets and JS to separate protected method, changing fronted framework will be pretty easy 

What about issues like #12588 then?

> I understand the theory. I was asking about its practical application for this particular case.

The original issue has 2 main tasks:

1) Exclude JQuery from the core classes.
2) Compose standalone JQuery extension, which usage will produce compatibility with the old Yii2-based code.

I do not consider this solution to be a final - it is the obvious way of solving task and should be improved.
But I need constractive ideas for it instead of arguing about the sense of the original issue.

Here I am trying to put things right, while someone may prefer to ignore the original design flaws and hide behind hacking solutions, like using CDN (proposed by @rob006) or put `provide` into `composer.jaon` (proposed by @cebe).

During the internal disscussions I have promised the core team, that I will show them the direction for splitting `yii2` package removing asset dependencies - so I did. But I can not enforce any solution on this matter.

**Comment by rob006** [comment]
> What about issues like #12588 then?

I don't get your question. Where is the problem? You have already solved this: https://github.com/klimov-paul/yii2/blob/ddcef80c72ae0babbf27d221d2b6a9ab65c1baaf/framework/web/View.php#L581-L606

**Comment by samdark** [comment]
This pull request touches many aspects of the framework at once. Let's address these one by one.

# yii\web\View and yii\jquery\View

| Q                      | A
| -------------------| -------------
| Is web only?     | Yes
| Inheritance issue? | Yes

I don't think `yii\jquery\View` is needed. Having  `document.addEventListener()` and  `window.addEventListener()` ways of doing things are enough.

## GridView

| Q                      | A
| -------------------| -------------
| Is web only?     | Yes
| Inheritance issue? | Yes

Looking at the code I see that most of it if not everything could be moved to a class which doesn't inherit from gridview itself:

```php
GridView::widget([]);
(new GridViewJavaScript($options))->register($this);

class GridViewJavaScript
{
    private $_options;

    public function __construct($options = [])
    {
        $this->_options = $options;
    }

    public function register($view)
    {
        GridViewAsset::register($view);
        // do something with options as it's currenly done in jquery version of gridview
    }
}
```

## Captcha

| Q                      | A
| -------------------| -------------
| Is web only?     | Yes

Could be moved into separate package as a whole and I plan doing that after we agree on how to deal with validators overall.
Widget part could be done similar to GridView.


## Validation

| Q                      | A
| -------------------| -------------
| Is web only?     | No (REST, CLI, code)
| Inheritance issue? | No

The builder solution applies to core validators only leaving extension validators as is i.e. containing both a method to validate value with PHP and a method to generate necessary validation JavaScript code.
It allows moving clientside part into jQuery package as planned but there are issues:

- Inconsistency between core and extensions is introduced.
- It's not clear what to do with extensions which are supposed to work both with and without jQuery.
- Validator while not implementing  clientValidateAttribute(), still has the methods defined. In the context of non-web usage it's not looking good.
- Big piece of code in a single method where each part isn't really connected with each other.

There's alternative solution @dynasource proposed in a short peer coding session today:

```php
abstract class AbstractValidator extends \yii\base\Object
{
    // validator could be modified by stacking more validators
    public $chainedValidators = [];

    /**
     * Executes the validation by running the current Validator and its chained Validators
     */
    final public function execute($value)
    {
        $this->validateValue($value);
        $this->runChainedValidators($value);       
    }

    /**
     * Validates a value by the current validator
     */
    abstract protected function validateValue($value) : bool;


    /**
     * Runs chained validators
     */
    private function runChainedValidators($value) : void
    {
        foreach($this->chainedValidators as $validator)
        {
             $validator = $this->createValidator($validator);
             $validator->execute($value);   
        }

    }

    /**
     * Creates Validator instance by array configuration
     */
    private function createValidator(array $config) : AbstractValidator
    {
         return Yii::createObject($config);   
    }
}

// base class for JavaScript part of the validation
abstract class AbstractJsValidator extends Validator
{
    protected function validateValue($value)
    {
        $js = $this->getJs();
        Yii::$app->view->registerJs($js); 
    }   

    abstract protected function getJs($value);

}

// --------------- third party extension --------------------


class CreditCardValidator extends AbstractValidator
{   
    // could be configured via DI container   
    public $chainedValidators = [
        ['class' => CreditCardJavaScriptExtraValidator::class]
    ];

    public function validate($value)
    {
        // actually validate credit card
    }
}


class CreditCardJavaScriptExtraValidator extends AbstractJsValidator
{

    protected function getJs($value)
    {
        // build JS necessary and return it           
    }   
}
```


## ActiveForm and ActiveField

| Q                      | A
| -------------------| -------------
| Is web only?     | Yes
| Inheritance issue? | Yes

Could be solved in the same manner as gridview.

**Comment by klimov-paul** [comment]
> I don't think yii\jquery\View is needed. Having document.addEventListener() and window.addEventListener() ways of doing things are enough.

I am not so sure. For `jQuery(window).on('load')` jQuery expression `window.addEventListener()` seems to be absolutely valid replacement, but for `jQuery(document).ready()` it is hard to say. It could happen `jQuery.ready()` is doing something more - otherwise it should not ever exitst.

> Looking at the code I see that most of it if not everything could be moved to a class which doesn't inherit from gridview itself

As I said [above](https://github.com/yiisoft/yii2/pull/13839#issuecomment-288995538) it could be a behavior. Then the view syntax will be following:

```php
<?= GridView::widget([
    'as jquery' => \yii\jquery\GridViewFileBehaviorr::class
]); ?>
```

Using absolutely standalone class will overcomplicate view code, besides client script composition usually requires data from the widget internal fields.

> Captcha could be moved into separate package. Widget part could be done similar to GridView.

Perhaps, but I am not sure it should.
Yii internal captcha has a poor strength (see #1100). It is fine to have it in the core as quick (mostly demo) solution.
Extracting it into a separated repo will increase developer's expectations from it and will require significant improvement of the CAPTCHA generation algorithm.
But in general I am not against the idea - it is fine to do so.

Anyway both `yii2-captcha` and `yii2-jquery` should be composed inside 2.1 branch before using GIT tree subsplit.
The main question is where JQuery dependent part of `Captcha` widget should be located in the end: at `yii2-captcha` or at `yii2-jquery`?

> Inconsistency between core and extensions is introduced.

At the present state there is no inconsistency: any existing code with custom validator, which `clientValidateAttribute()` returns not empty result,will function as before.
However it is likely to be working with JQuery only.

> Validator while not implementing clientValidateAttribute(), still has the methods defined. In the context of non-web usage it's not looking good.

`Validator::clientValidateAttribute()` can not be just removed, because of `InlineValidator` existance. 
Its client-side validation is determined by `Model::rules()` configuration and it is always developer-defined:
https://github.com/yiisoft/yii2/blob/master/framework/validators/InlineValidator.php#L78-L90

> Big piece of code in a single method where each part isn't really connected with each other.

Client validators can be extracted into a separated classes mapped to the existing server-side validators.
But they will be almost empty containing a 3-4 lines of code. Still I can adjust code in this way if you like.

> ActiveForm and ActiveField Could be solved in the same manner as gridview.

I am not so sure. Form validation is composed by both `ActiveForm` and `ActiveField`, which interacts together doing so.
It is hard to say if it can be just wrapped around.

**Comment by samdark** [comment]
> I am not so sure. For jQuery(window).on('load') jQuery expression window.addEventListener() seems to be absolutely valid replacement, but for jQuery(document).ready() it is hard to say. It could happen jQuery.ready() is doing something more - otherwise it should not ever exitst.

Well, yes. It's doing a bit more:

1. Providing polyfill for IE lower than 9.
2. Making sure callback will be executed even if code is executed after DOM ready.

Here's the code:

https://github.com/jquery/jquery/blob/master/src/core/ready-no-deferred.js

**Comment by samdark** [comment]
But that's not that helpful...

**Comment by samdark** [comment]
> As I said above it could be a behavior.

I like the idea. It could be configured via container as well...

**Comment by samdark** [comment]
> Anyway both yii2-captcha and yii2-jquery should be composed inside 2.1 branch before using GIT tree subsplit.

Why? I've successfully extracted https://github.com/yiisoft/yii2-maskedinput without doing a subsplit...

**Comment by samdark** [comment]
> The main question is where JQuery dependent part of Captcha widget should be located in the end: at yii2-captcha or at yii2-jquery?

At yii2-captcha but it should be a behavior or a separate class. Not a widget inherited from Captcha.

**Comment by samdark** [comment]
> Client validators can be extracted into a separated classes mapped to the existing server-side validators.
But they will be almost empty containing a 3-4 lines of code. Still I can adjust code in this way if you like.

I think it would make perfect sense.

**Comment by klimov-paul** [comment]
> Why? I've successfully extracted https://github.com/yiisoft/yii2-maskedinput without doing a subsplit

Subsplit is better as it will preserve the git history.

> At yii2-captcha but it should be a behavior or a separate class. Not a widget inherited from Captcha.

So it will, but my question still stands:

> where JQuery dependent part of Captcha widget should be located in the end: at `yii2-captcha` or at `yii2-jquery`?

If it will be located at `yii2-captcha` it will mean that this package will be dependent on `yii2-jquery` and can not be used without it. Thus it will be impossible to create a CAPTCHA solution without installing jQuery.
If it will be located `yii2-jquery` it will be dependent on `yii2-captcha`, thus you'll have to install captcha before using jQuery even if you do not need it.
The solid way to resolve it will be creating a 3rd package, which will depend on both `yii2-captcha` and `yii2-jquery`, but this may be an overcomplication.
Another solution can be putting JS-dependent part at one of `yii2-captcha` or `yii2-jquery` without explicit dependency ('suggets' composer section can be used for example), but this can be inconsistent.

**Comment by samdark** [comment]
About jQuery-related part of CAPTCHA, I prefer to rewrite JS part of it without using jQuery. yii.captcha.js isn't too complicated to do it.

**Comment by leandrogehlen** [comment]
My suggestion.

yii2-widgets (base widigets without js or css dependencies)
yii2-jquery (GridViewBehavior, and assets and capcha)

**Comment by samdark** [comment]
@leandrogehlen separating widgets isn't the goal of this pull request.

**Comment by leandrogehlen** [comment]
Ok, then , i think that the folder `grid` should be moved to `widgets` and removed assets dependencies from this classes, this way we can write a grid with `jQuery` and another with `Angular` , thus each extension can write a grid behavior with js library desired

**Comment by klimov-paul** [comment]
Solution has been refactored:

- Client script for `ActiveForm`, `GridView` and `Captcha` extracted into a behaviors.
- Client validation composition chnaged to use class map and special `ClientValidator` class
- `captcha` related code packed under `framework/captcha`

**Comment by samdark** [comment]
Looks good after quick glance on it. Need to check it in depth though.

**Comment by schmunk42** [comment]
I took a look at this and actually, the PR was looking good, at least no `assets` folder anymore in the framework.

I tried to merge against 2.1, but it looked to me like there were a lot of clientXyz functions added again?!

I think also that all the remaining asset-bundles should be removed from the core, but one step after the other...
bildschirmfoto 2017-09-05 um 16 43 19

Can this be made *mergeable* again?

**Comment by klimov-paul** [comment]
> Can this be made mergeable again?

At the present state I do not have time for this. I need to finish #14701 and underlying tasks, which depends on it. Otherwise, it will likely end in conflicts anyway.

Most likely, I will re-create this PR from scratch afterwards.

Since, at the present state, core team has no interest on reviewing this or attend to the underlying issue, it is not in hurry. Even if I *merge* this changeset, it will solves nothing until the `jquery` and `captcha` reositories will be split. While I do not have enough permissions to create a repositories under `yiisoft`, I am unable to to do it myself.
Without major core team suport this PR moves nowhere anyway.

**Comment by schmunk42** [comment]
> it will solves nothing until the jquery and captcha reositories will be split. While I do not have enough permissions to create a repositories under yiisoft, I am unable to to do it myself.

100% agreed! @samdark Could you create those repos for 2.1 (`yii2-jquery` and `yii2-captcha`; also `yii2-pjax`)? Those need to be moved out.

**Comment by samdark** [comment]
- https://github.com/yiisoft/yii2-jquery
- https://github.com/yiisoft/yii2-captcha

PJAX is to be removed so I haven't created a repo for it.

**Comment by klimov-paul** [comment]
Migrated to #14865

**Pull request closed without merging**